crates/gpui/src/platform/linux/wayland/window.rs

```diff
@@ -0,0 +1,348 @@
+        let mut wh = blade_rwh::WaylandWindowHandle::empty();
+        wh.surface = self.window;
+        wh.into()
+    }
+}
+
+unsafe impl HasRawDisplayHandle for RawWindow {
+    fn raw_display_handle(&self) -> RawDisplayHandle {
+        let mut dh = blade_rwh::WaylandDisplayHandle::empty();
+        dh.display = self.display;
+        dh.into()
+    }
+}
+
+impl WaylandWindowInner {
+    fn new(
+        conn: &Arc<wayland_client::Connection>,
+        wl_surf: &Arc<wl_surface::WlSurface>,
+        bounds: Bounds<i32>,
+    ) -> Self {
+        let raw = RawWindow {
+            window: wl_surf.id().as_ptr() as *mut _,
+            display: conn.backend().display_ptr() as *mut _,
+        };
+        let gpu = Arc::new(
+            unsafe {
+                gpu::Context::init_windowed(
+                    &raw,
+                    gpu::ContextDesc {
+                        validation: false,
+                        capture: false,
+                    },
+                )
+            }
+                .unwrap(),
+        );
+        let extent = gpu::Extent {
+            width: bounds.size.width as u32,
+            height: bounds.size.height as u32,
+            depth: 1,
+        };
+        Self {
+            renderer: BladeRenderer::new(gpu, extent),
+            bounds,
+        }
+    }
+}
+
+pub(crate) struct WaylandWindowState {
+    conn: Arc<wayland_client::Connection>,
+    inner: Mutex<WaylandWindowInner>,
+    pub(crate) callbacks: Mutex<Callbacks>,
+    pub(crate) surface: Arc<wl_surface::WlSurface>,
+    pub(crate) toplevel: Arc<xdg_toplevel::XdgToplevel>,
+}
+
+impl WaylandWindowState {
+    pub(crate) fn new(
+        conn: &Arc<wayland_client::Connection>,
+        wl_surf: Arc<wl_surface::WlSurface>,
+        toplevel: Arc<xdg_toplevel::XdgToplevel>,
+        options: WindowOptions,
+    ) -> Self {
+        if options.bounds == WindowBounds::Maximized {
+            toplevel.set_maximized();
+        } else if options.bounds == WindowBounds::Fullscreen {
+            toplevel.set_fullscreen(None);
+        }
+
+        let bounds: Bounds<i32> = match options.bounds {
+            WindowBounds::Fullscreen | WindowBounds::Maximized => Bounds {
+                origin: Point::default(),
+                size: Size {
+                    width: 500,
+                    height: 500,
+                }, //todo!(implement)
+            },
+            WindowBounds::Fixed(bounds) => bounds.map(|p| p.0 as i32),
+        };
+
+        Self {
+            conn: Arc::clone(conn),
+            surface: Arc::clone(&wl_surf),
+            inner: Mutex::new(WaylandWindowInner::new(&Arc::clone(conn), &wl_surf, bounds)),
+            callbacks: Mutex::new(Callbacks::default()),
+            toplevel,
+        }
+    }
+
+    pub fn update(&self) {
+        let mut cb = self.callbacks.lock();
+        if let Some(ref mut fun) = cb.request_frame {
+            fun();
+        }
```

Comment:
we shouldn't be locking the callbacks here this can cause a lock up when doing input if gpui requests to take the input_handler instead I think we need to take the request_frame and drop the lock like is done in macos. We probably should change this in x11 but that can be left till later
looks something like:
```rs
        let mut cb = self.callbacks.lock();
        if let Some(mut fun) = cb.request_frame.take() {
            drop(cb);
            fun();
            self.callbacks.lock().request_frame = Some(fun);
        }
```

Reply:
fixed